design-docs/007-basic-expiration.md

```diff
@@ -0,0 +1,561 @@
+              'git1': 'green',
+              'git2': 'orange',
+              'git3': 'brown',
+              'git4': 'gray',
+              'git5': '#888888',
+              'git6': '#888888',
+              'git7': '#888888'
+       } } }%%
+gitGraph
+
+
+    checkout main
+    commit type: HIGHLIGHT id: "0"
+
+    branch develop
+    checkout develop
+    commit id: "10"
+    commit id: "11"
+
+    checkout main
+    branch test
+    commit id: "9"
+
+    checkout main
+    branch qa
+    commit  type: REVERSE id: "1"
+    commit  type: REVERSE id: "2"
+
+    branch dangling-main
+    commit  type: REVERSE id: "4"
+    commit  type: REVERSE tag: "tag2" id: "5"
+    checkout qa
+
+
+
+    commit type: REVERSE  tag: "tag1" id: "3"
+    commit  type: REVERSE id: "6"
+    commit  type: REVERSE id: "7"
+    commit id: "8"
+
+    checkout main
+    commit  id: "12"
+    commit  id: "13"
+    commit  id: "14"
+```
+
+**Output after processing the `qa` branch**
+```mermaid
+---
+config:
+  gitGraph:
+    parallelCommits: true
+---
+
+%%{init: { 'logLevel': 'debug', 'theme': 'default' , 'themeVariables': {
+              'git0': 'blue',
+              'git1': 'green',
+              'git2': 'orange',
+              'git3': 'brown',
+              'git4': 'gray',
+              'git5': '#888888',
+              'git6': '#888888',
+              'git7': '#888888'
+       } } }%%
+gitGraph
+
+
+    checkout main
+    commit type: HIGHLIGHT id: "0"
+
+    branch develop
+    checkout develop
+    commit id: "10"
+    commit id: "11"
+
+    checkout main
+    branch test
+    commit id: "9"
+
+    checkout main
+    branch qa
+    commit id: "8"
+
+    checkout main
+    branch dangling-qa
+    commit  type: REVERSE id: "1"
+    commit  type: REVERSE id: "2"
+
+    branch dangling-main
+    commit  type: REVERSE id: "4"
+    commit  type: REVERSE id: "5" tag: "tag2"
+    checkout dangling-qa
+
+
+
+
+    commit type: REVERSE  tag: "ta2" id: "3"
+    commit  type: REVERSE id: "6"
+    commit  type: REVERSE id: "7"
+
+    checkout main
+    commit  id: "12"
+    commit  id: "13"
+    commit  id: "14"
+```
+
+Now snapshots `6` and `7` are dangling. They are ready to be garbage collected
+when GC runs. Snapshots `1` to `5` are still pointed by tags, but those
+tags are candidates for delete.
+
+When the algorithm processes tags `tag1` and `tag2` nothing new happens, because
+they already point to expired snapshots.
+
+The final result of the algorithm is a tree with a much wider shape. Garbage
+collection can already free two snapshots, and 5 more would be deleted if
+we delete the "expired tags". This process of deleting tags that are pointing to
+expired snapshots, could be optional part of the expiration process.
+
+
+
+## On timestamps
+
+There is a causality relationship between a snapshot and its parent: We know
+the parent was always written before the child, because it includes its id.
+But:
+
+* How do we enforce monotonicity of snapshot timestamps?
```

Comment:
How important is this causality really?

It seems to me that we only really have to know _the first snapshot to expire_ in a particular ancestry tree. After that, we can just traverse the ancestry, rather than explicitly comparing timestamps for each snapshot.

Reply:
Yes, it's just that the algorithm implementation is easier if we can assume every other ancestor is expired once we find the first expired snapshot. If there is clock drift, this could be no longer true.